Pass a stopper to RocksDB.Open() to properly close a rockdb instance

kkaneda opened the pull request:
NewInMem() calls Open() and adds an extra refcount to the engine, so we don't need call Open() again to keep rocksdb instances when restarting stores.

kkaneda committed:
Pass a stopper to RocksDB.Open to properly close a rockdb instance

We used to use a refcount, but there were tests that didn't close a rocksdb instance properly.
shutting down to deallocate a handle.

Start a gorountine from RocksDB.Open() to detect a leak.

bdarnell commented:
LGTM

So we've been leaking these in our tests the whole time? It would be nice if we had some way of detecting this, but I don't see any obvious way. We could start up a dummy goroutine in RocksDB.Open and let leaktest find that.

kkaneda commented:
Thanks for the review, @bdarnell . I see. Let me take a look.

kkaneda commented:
@bdarnell , could you take another look? There are many places where we called `NewInMem()` without closing, so I decided to pass a stopper to `NewInMem()` to properly close a rocksdb instance. The original change was reverted as we now need to call `Open` to keep a rocksdb instance when a store is closed.

As you suggested, changed `RocksDB.Open()` to start a gorountine to detect a leak.

Also fixed `TestResponseCacheInflight` to properly tear down the test. I got a panic from `engine._Cfunc_DBGet()`, and it seems the test didn't check the condition properly since `case <- c` can proceed multiple times when `c` has been closed.

tamird commented:
It's too bad this required so much plumbing, but thanks for doing it. LGTM

bdarnell commented:
It's unfortunate that we now have both the stopper and Open/Close. The interaction in client_test.go is especially strange: we create the Engine and the Store with the same Stopper, but we call Open() to allow the Engine to outlast the Store. Later we associate the Engine with a new Store, but there is no connection to that Store's Stopper. Other tests avoid the Open/Close dance but create a second Stopper to do so, and having two Stoppers in scope is confusing in its own way.

I think if we're going to require a Stopper for engine.NewInMem, we should go all the way and require it for NewRocksDB as well, and then get rid of the Open/Close refcount scheme completely (shouldn't be too hard since there aren't many call sites for NewRocksDB). Or we forget about moving everything to Stoppers and just use the goroutine leak detector to make sure that we're calling Close everywhere we need to (If you make the new goroutine a named method and add it to the blacklist in leaktest.AfterTest, leaktest will be much more helpful in finding the tests that need to change).

kkaneda commented:
Thanks, @tamird , @bdarnell . Agreed. It's awkward, and it's not very clear how these `Stopper`s interact.

Will first try the former approach (= get rid of Open/Close) and see how it looks like. Will also create a separate PR for `TestResponseCacheInflight` since it's not directly related.

kkaneda commented:
Could you take another look? Changed to pass `Stopper` to `RocksDB.Open`. The refcount was removed and `multiTestContext` uses another stopper (`engineStoppers`) to properly close engines.

bdarnell commented on storage/client_raft_test.go:
```diff
@@ -68,7 +68,9 @@ func TestStoreRecoverFromEngine(t *testing.T) {
 	manual := hlc.NewManualClock(0)
 	clock := hlc.NewClock(manual.UnixNano)
-	eng := engine.NewInMem(proto.Attributes{}, 1<<20)
+	stopper := stop.NewStopper()
```
Call the stoppers in this test and the next one `engineStopper` and `stopper` instead of `stopper` and `stopper2`.

kkaneda replied:
Sure, done!

bdarnell commented:
LGTM; thanks for cleaning this up.

tschottdorf commented:
sorry to revive this, but could this have to do with this recent change?

https://circleci.com/gh/tschottdorf/cockroach/1382

>
==================
WARNING: DATA RACE
Read by goroutine 25:
  github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).getProtoInternal()
      /go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:220 +0x16d
  github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).GetProto()
      /go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:210 +0x90
  github.com/cockroachdb/cockroach/storage/engine.InMem.GetProto()
      <autogenerated>:61 +0x9a
  github.com/cockroachdb/cockroach/storage/engine.MVCCGet()
      /go/src/github.com/cockroachdb/cockroach/storage/engine/mvcc.go:481 +0x300
  github.com/cockroachdb/cockroach/storage.(*Replica).Get()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:179 +0x125
  github.com/cockroachdb/cockroach/storage.(*Replica).executeCmd()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:69 +0x2e0f
  github.com/cockroachdb/cockroach/storage.(*Replica).executeBatch()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1174 +0x5fb
  github.com/cockroachdb/cockroach/storage.(*Replica).addReadOnlyCmd()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:783 +0x3b0
  github.com/cockroachdb/cockroach/storage.(*Replica).AddCmd()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:596 +0x786
  github.com/cockroachdb/cockroach/storage.TestRangeCommandQueue.func3()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:1112 +0x273

>
Previous write by goroutine 20:
  github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Close()
      /go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:138 +0x111
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).Stop()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:205 +0x15e
  github.com/cockroachdb/cockroach/storage.(*testContext).Stop()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:188 +0x45
  github.com/cockroachdb/cockroach/storage.TestRangeCommandQueue()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:1180 +0xc03
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

>
Goroutine 25 (running) created at:
  github.com/cockroachdb/cockroach/storage.TestRangeCommandQueue()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:1118 +0x807
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

>
Goroutine 20 (finished) created at:
  testing.RunTests()
      /tmp/workdir/go/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /tmp/workdir/go/src/testing/testing.go:494 +0xe4
  github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck()
      /go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x2e
  github.com/cockroachdb/cockroach/storage_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/storage/main_test.go:35 +0x2e
  main.main()
      github.com/cockroachdb/cockroach/storage/_test/_testmain.go:448 +0x209

>
fatal error: unexpected signal during runtime execution
[signal 0xb code=0x1 addr=0x0 pc=0x5bd03e]

>
runtime stack:
runtime.throw(0x15fb4a0, 0x2a)
	/tmp/workdir/go/src/runtime/panic.go:527 +0x90
runtime.sigpanic()
	/tmp/workdir/go/src/runtime/sigpanic_unix.go:12 +0x5a

>
goroutine 823 [syscall, locked to thread]:
runtime.cgocall(0x57a190, 0xc8200d6b80, 0xc800000000)
	/tmp/workdir/go/src/runtime/cgocall.go:120 +0xc2 fp=0xc8200d6b30 sp=0xc8200d6b08
github.com/cockroachdb/cockroach/storage/engine._Cfunc_DBGet(0x0, 0x0, 0xc8200ca048, 0x8, 0xc8202baf50, 0x0, 0x0)
	??:0 +0x7b fp=0xc8200d6b80 sp=0xc8200d6b30
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).getProtoInternal(0xc8200bab90, 0xc8200ca048, 0x8, 0x400, 0x7f8938d3c130, 0xc8200ca000, 0x0, 0x7f893a7a9000, 0x0, 0x0, ...)

tamird commented:
Yes, because the old implementation would just leak this thing.

tschottdorf commented:
it's a pretty drastic crash though. What probably happened is the `500ms` timeout that waits for the goroutine timing out. I'll move the goroutines into the stopper.

tamird commented:
I think that's a null pointer exception in C++ code.

tschottdorf commented:
yeah, because the goroutine tried to use the engine after it was destroyed. Will have a PR up soon.

kkaneda commented:
Hmm, sorry for the another issue..

tschottdorf commented:
@kkaneda your change helped catch a race. No need to be sorry :)
On Sep 23, 2015 05:51, "Kenji Kaneda" <notifications@github.com> wrote:

> Hmm, sorry for the another issue..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2521#issuecomment-142483377>
> .
>
